### PR TITLE
Workspace from devfile

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
Makes it possible to create a workspace from a devfile using a factory service

Add necessary functionality to the factory plugin.

create workspace from devfile

Solves issue https://github.com/eclipse/che/issues/12707

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

### CODE REVIEW

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-devfile-main.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
+/*********************************************************************
+ * Copyright (c) 2018 Red Hat, Inc.
```
Comment (olexii4) [inline extensions/eclipse-che-theia-plugin-ext/src/browser/che-devfile-main.ts]: 2019

File: extensions/eclipse-che-theia-plugin-ext/src/plugin/che-devfile.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+/*********************************************************************
+ * Copyright (c) 2018 Red Hat, Inc.
```
Comment (olexii4) [inline extensions/eclipse-che-theia-plugin-ext/src/plugin/che-devfile.ts]: 2019

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-devfile-main.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            resolve();
+        });
+    }
+
+    /**
+     * Opens window with URL to the factory.
+     */
+    protected async openFactoryWindow(uri: string): Promise<void> {
+        // do nothing for electron browser
+        if (isNative) {
```
Comment (olexii4) [inline extensions/eclipse-che-theia-plugin-ext/src/browser/che-devfile-main.ts]: We don't need to check it

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-devfile-main.ts (hunk 1 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/che-devfile-main.ts]: Done

File: extensions/eclipse-che-theia-plugin-ext/src/plugin/che-devfile.ts (hunk 2 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/plugin/che-devfile.ts]: Done

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-devfile-main.ts (hunk 3 above)
Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/che-devfile-main.ts]: Yep, indeed. Thanks

File: plugins/factory-plugin/src/devfile.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    constructor(private context: theia.PluginContext) {
+    }
+
+    init() {
+        this.context.subscriptions.push(theia.commands.registerCommand(
+            CREATE_WORKSPACE_COMMAND, uri => this.createWorkspace(uri))
+        );
+    }
+
+    async createWorkspace(uri: theia.Uri) {
```
Comment (benoitf) [inline plugins/factory-plugin/src/devfile.ts]: missing return type ?

File: plugins/factory-plugin/package.json [hunk 5]
```diff
@@ -53,5 +53,27 @@
             "json"
         ],
         "testURL": "http://localhost/"
+    },
+    "contributes": {
+        "menus": {
+            "editor/context": [
+                {
+                    "command": "factory-plugin.create-workspace",
+                    "when": "resourceFilename =~ /devfile/"
```
Comment (benoitf) [inline plugins/factory-plugin/package.json]: AFAIK now the devfile is named `.devfile.yml` no ?

Comment (vitaliy-guliy) [reply plugins/factory-plugin/package.json]: All names containing 'devfile' are suitable.
E.g. `devfile`, `.devfile`, `devfile.yaml` and others.
But this mask is not case sensitive, so `Devfile` will not be handled.

File: plugins/factory-plugin/src/devfile.ts (hunk 4 above)
Comment (vitaliy-guliy) [reply plugins/factory-plugin/src/devfile.ts]: void

Comment (vitaliy-guliy) [reply plugins/factory-plugin/src/devfile.ts]: It's a command handler. I suppose we can omit return type here. Isn't it?

Comment (benoitf) [reply plugins/factory-plugin/src/devfile.ts]: I don't know which tslint rules we're using but I was thinking it was mandatory

Comment (vitaliy-guliy) [reply plugins/factory-plugin/src/devfile.ts]: It's not a problem to add return type after adding corresponding rule for tslint.

**Pull request merged**